src/main/java/com/metamx/emitter/core/HttpPostEmitter.java

```diff
@@ -166,315 +190,344 @@ public void emit(Event event)
+      if (largeEventsToEmit.isEmpty()) {
+        return;
       }
-      catch (Throwable e) {
-        log.error(e, "Uncaught exception in EmittingRunnable.run()");
+      // Don't try to emit large events until exhaustion, to avoid starvation of "normal" batches, if large event
+      // posting rate is too high, though it should never happen in practice.
+      largeEventsToEmit.add(LARGE_EVENTS_STOP);
+      for (byte[] largeEvent; (largeEvent = largeEventsToEmit.poll()) != LARGE_EVENTS_STOP; ) {
+        emitLargeEvent(largeEvent);
       }
-
-      // Always reschedule, otherwise we all of a sudden don't emit anything.
-      exec.schedule(new EmittingRunnable(currVersion), config.getFlushMillis(), TimeUnit.MILLISECONDS);
     }
 
-    /**
-     * Serializes messages into a batch. Does not validate against maxBatchSize.
-     *
-     * @param messages list of JSON objects, one per message
-     * @return serialized JSON array
-     */
-    private byte[] serializeBatch(List<byte[]> messages)
+    private void emitLargeEvent(byte[] eventBytes)
     {
-      final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      byte[] buffer = acquireBuffer();
+      int bufferOffset = batchingStrategy.writeBatchStart(buffer);
+      System.arraycopy(eventBytes, 0, buffer, bufferOffset, eventBytes.length);
+      bufferOffset += eventBytes.length;
+      bufferOffset = batchingStrategy.writeBatchEnd(buffer, bufferOffset);
       try {
-        serializeTo(messages, baos);
-        return baos.toByteArray();
+        sendWithRetries(buffer, bufferOffset, 1);
       }
-      catch (IOException e) {
-        // There's no reason to have IOException in the signature of this method, since BAOS won't throw them.
-        throw Throwables.propagate(e);
+      finally {
+        buffersToReuse.add(buffer);
       }
     }
 
-    /**
-     * Serialize and compress (gzip) messages into a batch.
-     *
-     * @param messages list of JSON objects, one per message
-     * @return serialized and compressed JSON array
-     */
-    private byte[] serializeAndCompressBatch(List<byte[]> messages) throws IOException
+    private void sendWithRetries(final byte[] buffer, final int length, final int eventCount)
     {
-      final ByteArrayOutputStream baos = new ByteArrayOutputStream();
-      try (
-          final GZIPOutputStream gzos = new GZIPOutputStream(baos)
-      ) {
-        serializeTo(messages, gzos);
+      try {
+        RetryUtils.retry(
+            new Callable<Void>()
+            {
+              @Override
+              public Void call() throws Exception
+              {
+                send(buffer, length);
+                return null;
+              }
+            },
+            Predicates.<Throwable>alwaysTrue(),
+            MAX_SEND_RETRIES
+        );
+        totalEmittedEvents.addAndGet(eventCount);
+      }
+      catch (Exception e) {
+        log.error(e, "Failed to send events to url[%s]", config.getRecipientBaseUrl());
```

Comment:
Documentation to NonFatal explains it so that seems  `Error` + `InterruptedException`. We even don't catch Errors here, however it seems very strange to me that `InterruptedException` is considered Fatal.